File: soupServings/soupServings.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    double soupServings(int N) {
        if (N > 4800)
            return 1.0;
        vector<vector<double> > dp((N + 24) / 25 + 1, vector<double>((N + 24) / 25 + 1, -1.0));
        return dfs((N + 24) / 25, (N + 24) / 25, dp);
    }
    double dfs(int NA, int NB, vector<vector<double> >& dp) {
        if (NA <= 0 && NB <= 0)
            return 0.5;
        else if (NA <= 0)
            return 1;
        else if (NB <= 0)
            return 0;
        if (dp[NA][NB] > 0)
            return dp[NA][NB];
        dp[NA][NB] = 0.25 * (dfs(NA - 4, NB, dp) + dfs(NA - 3, NB - 1, dp) + dfs(NA - 2, NB - 2, dp) + dfs(NA - 1, NB - 3, dp));
        return dp[NA][NB];
    }
};

int main() {
  Solution s;
  cout << s.soupServings(300) << endl;
  return 0;
}

File: soupServings/soupServings.java
class Solution {
    public double soupServings(int n) {
        if (n > 4800) return 1.0;
        n = (n + 24) / 25;
        double[][] dp = new double[n + 1][n + 1];
        dp[0][0] = 0.5;
        for (int j = 1; j <= n; j++) dp[0][j] = 1;
        int[][] serve = {{4, 0}, {3, 1}, {2, 2}, {1, 3}};
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                for (int[] s : serve) {
                    int prevI = i >= s[0] ? i - s[0] : 0;
                    int prevJ = j >= s[1] ? j - s[1] : 0;
                    dp[i][j] += 0.25 * dp[prevI][prevJ];
                }
            }
        }
        return dp[n][n];
    }
}
